Under memory pressure, non active SourceBuffers should be completely emptied

jyavenard opened the pull request:
#### 6ff5566a514b479065e34a377acd3f6a00173417
<pre>
Under memory pressure, non active SourceBuffers should be completely emptied
<a href="https://bugs.webkit.org/show_bug.cgi?id=252957">https://bugs.webkit.org/show_bug.cgi?id=252957</a>
rdar://105940463

Reviewed by Youenn Fablet.

Fully clear all track buffers of a non active source buffer as it&apos;s not
currently playing.

* LayoutTests/media/media-source/media-managedmse-memorypressure-inactive-expected.txt: Added.
* LayoutTests/media/media-source/media-managedmse-memorypressure-inactive.html: Added.
* LayoutTests/platform/glib/TestExpectations: Memory Pressure handler isn&apos;t set on these platforms.
* Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
(WebCore::SourceBufferPrivate::clearTrackBuffers):
(WebCore::SourceBufferPrivate::memoryPressure):
(WebCore::SourceBufferPrivate::removeCodedFrames): Fly by fix, ensure m_client is set.
* Source/WebCore/platform/graphics/SourceBufferPrivate.h:
* Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.cpp:
(WebKit::SourceBufferPrivateRemote::clearTrackBuffers):
* Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.h:

Canonical link: <a href="https://commits.webkit.org/260870@main">https://commits.webkit.org/260870@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b546c8573047b1cf9fca9cc11ad15a404bfc9937

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/109759 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/18875 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/42497 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/1230 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/118878 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/113710 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/20345 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/10072 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/102027 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/115508 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/15145 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/98376 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/43368 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/97114 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/30029 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/85160 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/11601 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/31371 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/12253 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/8311 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/17622 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/50980 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7536 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/14003 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
```diff
@@ -189,10 +189,18 @@ void SourceBufferPrivate::seekToTime(const MediaTime& time)
 }
 
-void SourceBufferPrivate::clearTrackBuffers()
+void SourceBufferPrivate::clearTrackBuffers(bool reportClient)
```
s/reportClient/shouldReportToClient?

jyavenard replied:
done

youennf commented on Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
```diff
@@ -189,10 +189,18 @@ void SourceBufferPrivate::seekToTime(const MediaTime& time)
+    updateBufferedFromTrackBuffers(true);
+    updateHighestPresentationTimestamp();
+
+    m_client->sourceBufferPrivateReportExtraMemoryCost(totalTrackBufferSizeInBytes());
```
Are we sure m_client is always never null?

jyavenard replied:
sounds like it could be if called before setClient is called, there are some missing check too, will do some flyby

jyavenard committed:
Under memory pressure, non active SourceBuffers should be completely emptied
https://bugs.webkit.org/show_bug.cgi?id=252957
rdar://105940463

Reviewed by Youenn Fablet.

Fully clear all track buffers of a non active source buffer as it's not
currently playing.

* LayoutTests/media/media-source/media-managedmse-memorypressure-inactive-expected.txt: Added.
* LayoutTests/media/media-source/media-managedmse-memorypressure-inactive.html: Added.
* LayoutTests/platform/glib/TestExpectations: Memory Pressure handler isn't set on these platforms.
* Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
(WebCore::SourceBufferPrivate::clearTrackBuffers):
(WebCore::SourceBufferPrivate::memoryPressure):
(WebCore::SourceBufferPrivate::removeCodedFrames): Fly by fix, ensure m_client is set.
* Source/WebCore/platform/graphics/SourceBufferPrivate.h:
* Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.cpp:
(WebKit::SourceBufferPrivateRemote::clearTrackBuffers):
* Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.h:

Canonical link: https://commits.webkit.org/260870@main
